torch/onnx/_internal/fx/onnxfunction_dispatcher.py

```diff
@@ -431,38 +447,84 @@ class _OnnxSchemaChecker:
 
     It provides methods to check for input compatibility based on the OpSchema. It also
     provides a matching score to indicate how well the OpSchema matches the input and
-    kwargs types.
+    kwargs types. A function will be evaluated as perfect match, nearest match eligible,
+    or no match.
 
-    There are three types of ONNX overloads in torchlib:
+    Here are some common examples in categories:
 
-    1. Different types: Caused by the difference between the ONNX spec and PyTorch.The
-        matching system finds the correct one.
+    1. [NOTE: Perfect match]: The number of inputs and attributes are exactly the same as
+        the OpSchema. The types of inputs and attributes are exactly the same as the
+        OpSchema.
 
         ```python
-        @torch_op("aten::mul")
-        def aten_mul(self: TReal, other: TReal) -> TReal:
+        inputs = (Tensor[2, 3], Tensor[2, 3])
+        attributes = {"alpha": 1.0}
+
+        @torch_op("aten::op")
+        def aten_op(self: TReal, other: TReal, alpha: float = 1) -> TReal:
             ...
 
-        @torch_op("aten::mul")
-        def aten_mul_bool(self: BOOL, other: BOOL) -> BOOL:
+        ```
+        Result: Perfect match.
+
+    2. [NOTE: Optional input]: The dispatcher recognizes optional inputs. However,
+        the input can't be ignored. None must be provided.
+
+        ```python
+        inputs = (Tensor([2, 3]), None)
+        attributes = {}
+
+        aten_op(X: TTensor, Y: Optional[INT64]):
             ...
-    ```
+        ```
+        Result: Perfect match.
+        Real example: `aten::convolution`.
 
-    2. Optional dtype: dtype could be "unprovided". The difference from 2 is that dtype
-        would not be None.
+    3. [NOTE: Different attributes]: If an attribute is provided with value, it's
+        a must to match the attribute in function signature.
+        ```python
+        inputs = (Tensor([2, 3]),)
+        attributes = {"a":1, "b":2}
 
+        aten_op(X: TTensor, a: int):
+            ...
+        ```
+        Result: No match.
+        Real example: `aten::div` vs `aten::div.Tensor_mode`.
+
+    4. [NOTE: Default attributes]: Default attribute will fill in the value into
+        inputs/attributes.
         ```python
-        @torch_op("aten::new_full")
-        def aten_new_full(self: TTensor, size: INT64, fill_value: TTensor) -> TTensor:
+        inputs = (Tensor([2, 3]),)
+        attributes = {}
+
+        aten_op(X: TTensor, a: int = 3):
             ...
+        ```
+        Result: Perfect match.
+        Real example: `aten::clone`
+
+    5. [NOTE: Ignore attribute with None value]: The attributes with None value
```

Comment:
I guess this is due to the ONNXScript limitation I typed above? Cannot define optional attribute?

Reply:
I think it's something about AttributeProto, but yes, we don't have Optional attribute at the time, and this assumption is acknowledged by OpSchemaWrapper.

Semantically, besides dtype, another goal of creating ONNX variants is to overload the cases with different attributes, because they are usaully composed by different set of ops. Thus, attribute with None value simply means the function doesn't need this attribute.

cc @justinchuby

Reply:
I will need to double check with Rama on optional attributes. IIRC we either provide them in the graph or not provide them in the graph. There is not a "None" as the attribute value. In the torchscript builder, we omit None attributes whenever we see them.